### ISSUE TITLE
training acc and ppl is unstable compareed with onmt-py 0.3.1

### ISSUE THREAD

**Issue by sys1874** [OpenNMT/OpenNMT-py#1025]
**version 0.4.1**    pytorch 0.4.1  torchtext 0.4.0
2018-11-01 18:44:26,801 INFO] Step 31150/100000000000; acc:  58.42; ppl:  7.46; xent: 2.01; lr: 0.00030; 12686/437 tok/s;  23345 sec
[2018-11-01 18:45:01,898 INFO] Step 31200/100000000000; acc:  33.19; ppl: 45.57; xent: 3.82; lr: 0.00030; 12352/305 tok/s;  23380 sec
[2018-11-01 18:45:34,106 INFO] Step 31250/100000000000; acc:  60.73; ppl:  8.19; xent: 2.10; lr: 0.00030; 10570/353 tok/s;  23412 sec
[2018-11-01 18:46:09,160 INFO] Step 31300/100000000000; acc:  42.15; ppl: 27.04; xent: 3.30; lr: 0.00030; 9120/556 tok/s;  23447 sec
[2018-11-01 18:46:42,420 INFO] Step 31350/100000000000; acc:  65.47; ppl:  9.08; xent: 2.21; lr: 0.00030; 11366/374 tok/s;  23480 sec
[2018-11-01 18:47:16,870 INFO] Step 31400/100000000000; acc:  49.25; ppl: 16.78; xent: 2.82; lr: 0.00030; 11377/462 tok/s;  23515 sec
[2018-11-01 18:47:49,836 INFO] Step 31450/100000000000; acc:  50.92; ppl: 13.50; xent: 2.60; lr: 0.00030; 9679/451 tok/s;  23548 sec
[2018-11-01 18:48:25,120 INFO] Step 31500/100000000000; acc:  36.64; ppl: 38.30; xent: 3.65; lr: 0.00030; 11499/270 tok/s;  23583 sec
[2018-11-01 18:48:59,652 INFO] Step 31550/100000000000; acc:  40.85; ppl: 25.37; xent: 3.23; lr: 0.00030; 10975/284 tok/s;  23617 sec
[2018-11-01 18:49:32,964 INFO] Step 31600/100000000000; acc:  45.15; ppl: 20.61; xent: 3.03; lr: 0.00030; 12060/419 tok/s;  23651 sec
[2018-11-01 18:50:06,834 INFO] Step 31650/100000000000; acc:  48.68; ppl: 13.91; xent: 2.63; lr: 0.00030; 10962/343 tok/s;  23685 sec
[2018-11-01 18:50:41,490 INFO] Step 31700/100000000000; acc:  35.84; ppl: 31.35; xent: 3.45; lr: 0.00030; 9934/333 tok/s;  23719 sec
**In epoch 1, the training acc and ppl are unstable during training.**

**version 0.3.1**    pytorch 0.3.1   torchtext 0.2.3
Epoch  1,  5300/10379; acc:  35.76; ppl:  44.42; xent:   3.79; 17722 src tok/s; 576 tgt tok/s;   2760 s elapsed
Epoch  1,  5350/10379; acc:  37.86; ppl:  39.32; xent:   3.67; 17763 src tok/s; 600 tgt tok/s;   2785 s elapsed
Epoch  1,  5400/10379; acc:  36.57; ppl:  42.29; xent:   3.74; 17811 src tok/s; 580 tgt tok/s;   2811 s elapsed
Epoch  1,  5450/10379; acc:  35.60; ppl:  45.68; xent:   3.82; 17704 src tok/s; 594 tgt tok/s;   2836 s elapsed
Epoch  1,  5500/10379; acc:  36.51; ppl:  42.95; xent:   3.76; 17501 src tok/s; 572 tgt tok/s;   2863 s elapsed
Epoch  1,  5550/10379; acc:  38.58; ppl:  38.54; xent:   3.65; 17574 src tok/s; 584 tgt tok/s;   2889 s elapsed
Epoch  1,  5600/10379; acc:  37.04; ppl:  41.57; xent:   3.73; 17173 src tok/s; 580 tgt tok/s;   2915 s elapsed
Epoch  1,  5650/10379; acc:  37.23; ppl:  40.75; xent:   3.71; 17119 src tok/s; 583 tgt tok/s;   2941 s elapsed
Epoch  1,  5700/10379; acc:  36.39; ppl:  42.33; xent:   3.75; 17346 src tok/s; 565 tgt tok/s;   2968 s elapsed
Epoch  1,  5750/10379; acc:  37.17; ppl:  40.30; xent:   3.70; 17542 src tok/s; 568 tgt tok/s;   2994 s elapsed
Epoch  1,  5800/10379; acc:  38.80; ppl:  36.28; xent:   3.59; 17180 src tok/s; 584 tgt tok/s;   3021 s elapsed
Epoch  1,  5850/10379; acc:  38.44; ppl:  37.27; xent:   3.62; 17833 src tok/s; 571 tgt tok/s;   3046 s elapsed
Epoch  1,  5900/10379; acc:  37.25; ppl:  39.94; xent:   3.69; 17150 src tok/s; 579 tgt tok/s;   3073 s elapsed
Epoch  1,  5950/10379; acc:  38.91; ppl:  35.97; xent:   3.58; 17533 src tok/s; 597 tgt tok/s;   3099 s elapsed
Epoch  1,  6000/10379; acc:  38.37; ppl:  39.24; xent:   3.67; 17587 src tok/s; 557 tgt tok/s;   3125 s elapsed
Epoch  1,  6050/10379; acc:  37.58; ppl:  39.91; xent:   3.69; 17490 src tok/s; 588 tgt tok/s;   3152 s elapsed
**In epoch 1, the training acc and ppl are improved stably during training.**

Have anyone met the same problem?

**Comment by vince62s** [comment]
post your command lines and and describe the task.

**Comment by sys1874** [comment]
I use onmt to run a summary task.
**For onmt 0.4.1** 
python train.py -save_model data/baseline  \
                -data /data1/data          \
                -copy_attn \
                -global_attention mlp \
                -word_vec_size 300 \
                -rnn_size 1024 \
                -layers 2 \
		-accum_count 1 \
                -encoder_type brnn \
                -max_grad_norm 5 \
                -dropout 0.5\
                -batch_size 20 \
                -optim adam \
                -learning_rate 0.0003 \
                -reuse_copy_attn \
                -copy_loss_by_seqlength \
		-param_init_glorot   \
                -bridge \
                -seed 777 \
                -world_size 1 \
                -gpu_ranks 7 \
		-valid_steps 3000 \
		-valid_batch_size 10 \
		-train_steps 100000000000   

**For onmt 0.3.1** 
 python  train.py  -gpuid 1 \
                            -learning_rate 0.0003  \
                           -batch_size 24 \
                           -save_model data/model/baseline  \
                           -data data1/data   \
                           -global_attention mlp  \
                           -word_vec_size 300 \
                           -rnn_size 1024 \
                           -layers 2 \
                           -encoder_type brnn \
                           -epochs 20 \
                           -max_grad_norm 5 \
                           -dropout 0.5 \
                           -optim adam  \
                          -coverage_attn  \
                          -copy_attn \
                          -reuse_copy_attn  \
                          -bridge \
                          -seed 777 \
                          -valid_batch_size  5  \
                          -param_init_glorot \   

**I have try kinds of settings for onmt 0.4.1** 
Using -copy_loss_by_seqlength  or not 
Using -coverage_attn or not
Using -param_init_glorot or not 

The model of onmt 0.4.1 still converges, but it's train acc and ppl looks unstable than onmt 0.3.1 .

**Comment by sys1874** [comment]
I run my program in the server. And there is another strange phenomenon. 
When there is very little free memory，runing onmt 0.4.1 will cause that the src/tgt numbers of each batch is unstable:

**little free memory :**
[2018-11-01 08:44:50,931 INFO] Step 100/100000000000; acc:   9.22; ppl: 175.47; xent: 5.17; lr: 0.00030; 5714/198 tok/s;    165 sec
[2018-11-01 08:46:13,737 INFO] Step 150/100000000000; acc:  11.07; ppl: 171.55; xent: 5.14; lr: 0.00030; 9505/326 tok/s;    248 sec
[2018-11-01 08:49:17,800 INFO] Step 200/100000000000; acc:   9.12; ppl: 177.93; xent: 5.18; lr: 0.00030; 1134/ 43 tok/s;    432 sec
[2018-11-01 08:50:53,459 INFO] Step 250/100000000000; acc:  11.02; ppl: 161.21; xent: 5.08; lr: 0.00030; 12395/377 tok/s;    528 sec
[2018-11-01 08:54:12,583 INFO] Step 300/100000000000; acc:  14.09; ppl: 237.90; xent: 5.47; lr: 0.00030; 9574/240 tok/s;    727 sec
[2018-11-01 08:56:13,830 INFO] Step 350/100000000000; acc:  15.65; ppl: 144.70; xent: 4.97; lr: 0.00030; 2998/130 tok/s;    848 sec
[2018-11-01 08:57:44,758 INFO] Step 400/100000000000; acc:  13.97; ppl: 261.14; xent: 5.57; lr: 0.00030; 5652/207 tok/s;    939 sec
[2018-11-01 08:58:54,215 INFO] Step 450/100000000000; acc:  15.14; ppl: 203.71; xent: 5.32; lr: 0.00030; 10208/325 tok/s;   1009 sec
[2018-11-01 09:00:48,777 INFO] Step 500/100000000000; acc:  14.80; ppl: 133.67; xent: 4.90; lr: 0.00030; 1099/ 31 tok/s;   1123 sec
[2018-11-01 09:02:53,164 INFO] Step 550/100000000000; acc:  15.90; ppl: 180.93; xent: 5.20; lr: 0.00030; 9007/487 tok/s;   1247 sec
[2018-11-01 09:04:41,289 INFO] Step 600/100000000000; acc:  19.00; ppl: 100.36; xent: 4.61; lr: 0.00030; 6617/224 tok/s;   1356 sec
[2018-11-01 09:06:23,462 INFO] Step 650/100000000000; acc:  17.20; ppl: 170.46; xent: 5.14; lr: 0.00030; 9997/284 tok/s;   1458 sec
[2018-11-01 09:08:54,959 INFO] Step 700/100000000000; acc:  24.17; ppl: 103.79; xent: 4.64; lr: 0.00030; 2935/ 73 tok/s;   1609 sec
[2018-11-01 09:10:34,934 INFO] Step 750/100000000000; acc:  22.47; ppl: 112.52; xent: 4.72; lr: 0.00030; 10151/312 tok/s;   1709 sec
[2018-11-01 09:12:36,624 INFO] Step 800/100000000000; acc:  16.34; ppl: 108.21; xent: 4.68; lr: 0.00030; 1889/ 81 tok/s;   1831 sec

**Sufficient free memory :**
[2018-11-01 12:16:29,224 INFO] Step 100/100000000000; acc:   9.22; ppl: 175.47; xent: 5.17; lr: 0.00030; 11735/407 tok/s;     67 sec
[2018-11-01 12:17:01,959 INFO] Step 150/100000000000; acc:  11.07; ppl: 171.55; xent: 5.14; lr: 0.00030; 11916/409 tok/s;    100 sec
[2018-11-01 12:17:36,296 INFO] Step 200/100000000000; acc:   9.12; ppl: 177.93; xent: 5.18; lr: 0.00030; 10295/395 tok/s;    134 sec
[2018-11-01 12:18:10,017 INFO] Step 250/100000000000; acc:  11.02; ppl: 161.21; xent: 5.08; lr: 0.00030; 12547/382 tok/s;    168 sec
[2018-11-01 12:18:41,934 INFO] Step 300/100000000000; acc:  14.09; ppl: 237.90; xent: 5.47; lr: 0.00030; 12437/311 tok/s;    200 sec
[2018-11-01 12:19:16,058 INFO] Step 350/100000000000; acc:  15.65; ppl: 144.70; xent: 4.97; lr: 0.00030; 10379/449 tok/s;    234 sec
[2018-11-01 12:19:48,284 INFO] Step 400/100000000000; acc:  13.97; ppl: 261.14; xent: 5.57; lr: 0.00030; 11054/405 tok/s;    266 sec
[2018-11-01 12:20:22,614 INFO] Step 450/100000000000; acc:  15.14; ppl: 203.71; xent: 5.32; lr: 0.00030; 10988/349 tok/s;    300 sec
[2018-11-01 12:20:55,957 INFO] Step 500/100000000000; acc:  14.80; ppl: 133.67; xent: 4.90; lr: 0.00030; 11203/319 tok/s;    334 sec
[2018-11-01 12:21:28,548 INFO] Step 550/100000000000; acc:  15.90; ppl: 180.93; xent: 5.20; lr: 0.00030; 9751/527 tok/s;    366 sec
[2018-11-01 12:22:02,142 INFO] Step 600/100000000000; acc:  19.00; ppl: 100.36; xent: 4.61; lr: 0.00030; 12319/417 tok/s;    400 sec
[2018-11-01 12:22:36,267 INFO] Step 650/100000000000; acc:  17.20; ppl: 170.46; xent: 5.14; lr: 0.00030; 12686/360 tok/s;    434 sec
[2018-11-01 12:23:08,954 INFO] Step 700/100000000000; acc:  24.17; ppl: 103.79; xent: 4.64; lr: 0.00030; 12479/312 tok/s;    467 sec

**Comment by vince62s** [comment]
-world_size 1 -gpu_ranks 7 is not correct. it has to start at 0. if you need GPU 7 only available use CUDA_VISIBLE_DEVICES
You don't have the same options, if you want to compare make them exactly the same.
what language is it? (in case there could be some unicode inconsistencies)
dropout = 0.5 is this on purpose ? it looks way too high
On your second post at 700 PPL and ACC are exactly the same, only timing differs because of swapping I guess.

**Comment by sys1874** [comment]
Oh my god, I find the problem finally. It almost makes me crazy  !!!!
In report_manager.py  line 75:
**def report_training()**  will return a new report_sate no matter weather the step % 50 ==0 . 
So,  it just report the information from the 50th batch, which cause the result looks like unstable.

change the code as follow will solve the problem.
`    

        if step % self.report_every == 0:
            self._report_training(
                step, num_steps, learning_rate, report_stats)
            self.progress_step += 1
            return onmt.utils.Statistics()
        else:
            return report_stats`

**Comment by vince62s** [comment]
What you mean by "unstable" is that you are looking for cumulative stats vs instant stats.
But the current behavior is actually done on purpose.
@pltrdy tell me if I am wrong.

**Comment by sys1874** [comment]
Yeah, the training acc will improve as smoothly as opnmt 0.3.1, which looks more comfortable.

**Comment by pltrdy** [comment]
Nice one, thanks! 

Could you open a PR for this?

**Comment by sys1874** [comment]
I have opened a PR for it. 😸
